### PR TITLE
feat: create user endpoint (TBC-17)

### PR DESCRIPTION
**Pull request by watermalon3**
### PR Checklist

Please check for the following requirements:

- [x] I have reviewed this PR's changes at least once
- [ ] The squashed commit message explains exactly what you changed, added or deleted

#### Changes

- Replace with changes made

##### Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

#### What is the current behavior?

- [Ticket](https://travelingbookclub.atlassian.net/browse/tbc-17)
- [ ] Another developer ran through the acceptance criteria locally (bugs fixes and features)
- [ ] I ran through the acceptance criteria locally (bugs fixes and features)
- [ ] I have added tests for meeting the acceptance criteria (bugs fixes and features)

### CODE REVIEW

**Comment by watermalon3** [comment]
Screenshot 2023-05-23 at 12 18 49 PM

**Comment by watermalon3** [comment]
Screenshot 2023-05-23 at 12 43 07 PM

**Comment by watermalon3** [comment]
Screenshot 2023-05-23 at 1 35 48 PM

**Comment by watermalon3** [comment]
Screenshot 2023-05-23 at 1 36 12 PM

when not all information is present

File: backend/controllers/Auth.js
Comment (watermalon3) [inline backend/controllers/Auth.js:1]: this is just checking if we have everything that is required for the user schema and if so we create one and then send back that object. we can add a jwt token in here very easily if that is how we want to authenticate

File: backend/models/User.js
Comment (watermalon3) [inline backend/models/User.js:1]: i feel like we need a password to be part of this what do you think?

File: backend/controllers/Auth.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+const router = require("express").Router();
+const User = require("../models/User");
+
+router.post("/register", async (req, res) => {
+  try {
+    const { firstName, lastName, email } = req.body;
+    if (!firstName || !lastName || !email) {
+      throw new Error("The user schema is not complete");
+    }
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]: The error should be specific to what is missing for the front end. 
ex. If the last name isn't provided, then it should send a message back, "last name is required"

File: backend/controllers/Auth.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    const newUser = new User({
+      firstName,
+      lastName,
+      email,
+    });
+
+    await newUser.save();
+
+    res.status(200).json({
+      message: "User successfully create",
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]:
```suggestion
      message: "User successfully created",
```

File: backend/controllers/Auth.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    }
+    const newUser = new User({
+      firstName,
+      lastName,
+      email,
+    });
+
+    await newUser.save();
+
+    res.status(200).json({
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]: This should be a `201` status code
```
201: “Created.” The server has fulfilled the browser’s request, and as a result, has created a new resource.
```
https://kinsta.com/blog/http-status-codes/

File: backend/server.js [hunk 4]
```diff
@@ -1,13 +1,16 @@
 const express = require("express");
+const mongoose = require("mongoose");
 const app = express();
 const cors = require("cors");
 require("dotenv").config();
 app.use(cors());
 app.use(express.json());
 
 const source = process.env.ATLAS_CONNECTION;
+const PORT = 4000;
```
Comment (ohnomalone) [inline backend/server.js:10]: Is this the report to run locally?

File: backend/models/User.js
Comment (ohnomalone) [reply backend/models/User.js:1]: Where did you put it on your other app? Something to think about, in places the user is requested by the front-end, we don't want the password returned.

File: backend/controllers/Auth.js
Comment (ohnomalone) [reply backend/controllers/Auth.js:1]: The file naming convention, are they supposed to be capitalized?

**Review by ohnomalone — CHANGES_REQUESTED** [review]
This is looking great! A few comments and updates then we'll be on our way.

File: backend/controllers/Auth.js (hunk 1 above)
Comment (ohnomalone) [inline backend/controllers/Auth.js]: To add in email validation we can do 

``` 
const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(email)) {
      throw new Error("Invalid email address");
    }
```
a regular expression (emailRegex) is used to validate the email format. The regular expression pattern /^[^\s@]+@[^\s@]+\.[^\s@]+$/ checks if the email address has the basic structure of <username>@<domain>.<tld>

File: backend/controllers/Auth.js
Comment (watermalon3) [reply backend/controllers/Auth.js:1]: no I just updated it

Comment (watermalon3) [reply backend/controllers/Auth.js:1]: except for the models they should be capitalize

File: backend/models/User.js
Comment (watermalon3) [reply backend/models/User.js:1]: it was part of the other app but used encryption and a secret key so you couldn't use it but I feel like there is a better system

File: backend/controllers/Auth.js (hunk 1 above)
Comment (watermalon3) [reply backend/controllers/Auth.js]: Screenshot 2023-05-24 at 12 04 05 PM

File: backend/controllers/Auth.js (hunk 2 above)
Comment (watermalon3) [reply backend/controllers/Auth.js]: done

File: backend/server.js (hunk 4 above)
Comment (watermalon3) [reply backend/server.js:10]: yeah this is just for it to run locally and know what port it is using

File: backend/controllers/Auth.js (hunk 1 above)
Comment (watermalon3) [reply backend/controllers/Auth.js]: I could build that into the modal itself or should it be in auth?

File: backend/controllers/Auth.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    const { firstName, lastName, email } = req.body;
+    if (!firstName) {
+      throw new Error("the first name is missing");
+    }
+    if (!lastName) {
+      throw new Error("last name missing");
+    }
+    if (!email) {
+      throw new Error("email missing");
+    }
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]:
```suggestion

enum fields = {
firstName = "first name",
lastName = "lastName",
email = "email"
}
    if (!firstName || !lastName || !email ) {
    const missingFields = []
    Object.keys(req.body).forEach((field) => {
    if (!field) {
    missingFields.push(fields[field])
    })
      throw new Error(`${missingFields.join()} is missing");
    }
```

File: backend/controllers/Auth.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+      email,
+    });
+
+    await newUser.save();
+    const { _id } = newUser;
+    res.status(201).json({
+      message: "User successfully created",
+      firstName,
+      lastName,
+      _id,
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]: so what this does is create a key `_id` which isn't a practice we have on the front end. A way that works better on the fe is `id: newUser._id` so on the front end the object just contains normal keys.

File: backend/controllers/Auth.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  firstName: "first name",
+  lastName: "last name",
+  email: "email",
+};
+router.post("/register", async (req, res) => {
+  try {
+    const { firstName, lastName, email } = req.body;
+    if (!firstName || !lastName || !email) {
+      const missingFields = [];
+      console.log(Object.keys(req.body), "req.body");
```
Comment (ohnomalone) [inline backend/controllers/Auth.js]: All the logs should be removed

File: backend/models/User.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+const mongoose = require("mongoose");
+const User = new mongoose.Schema({
+  firstName: {
+    type: String,
+    required: true,
+  },
+  lastName: {
+    type: String,
+    required: true,
```
Comment (ohnomalone) [inline backend/models/User.js:9]: With these being required true, what is the error message sent back? I think we might be able to remove all of the validations in the `Auth.js` file and customize any error response back to the frontend.

**Review by ohnomalone — CHANGES_REQUESTED** [review]
Thanks for working on this! The biggest thing to research and use is built in validation, after we tell the service a field is valid, it takes care of checking.

File: backend/controllers/Auth.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  email: "email",
+};
+router.post("/register", async (req, res) => {
+  try {
+    const { firstName, lastName, email } = req.body;
+    const newUser = new User({
+      firstName,
+      lastName,
+      email,
+    });
```
Comment (ohnomalone) [inline backend/controllers/Auth.js:16]: Could we simplify this further? Maybe:

```suggestion
    const newUser = new User(req.body);
```

File: backend/controllers/Auth.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    await newUser.save();
+    res.status(201).json({
+      message: "User successfully created",
+      firstName,
+      lastName,
+      id: newUser._id,
```
Comment (ohnomalone) [inline backend/controllers/Auth.js:23]: We want to return the email as well.

**Review by ohnomalone — CHANGES_REQUESTED** [review]
Loving all the progress! 
Left a couple of thoughts.

**Review by ohnomalone — APPROVED** [review]
Thanks for all the hard work on this one! 👍

**Pull request merged**